### DISCUSSION TITLE
Avatar Carryables

### DISCUSSION THREAD

**Discussion by dpogue** [H-uru/Plasma#974] [Ideas & Proposals]
Pulling the comments from https://github.com/H-uru/Plasma/pull/972 into a discussion so they don't get lost. I'm keeping this simple and nowhere near the scope of a full cross-Age inventory system.

@Hoikas says:
> [...]  I think it would be better to have a more formalized inventory system for attaching objects to the avatar (this system probably involves avatar.sdl) with the ability to anchor objects to the left hip or off-hand built-in, for example. That way, we discourage bespoke solutions in each Age that results in an increased maintenance burden.

This seems (at face value) like the SDL side could be as easy as adding 2 `PLKEY` variables [here](https://github.com/H-uru/Plasma/blob/b00499f3e4149f8937094d1dfdfb7f8dc82596d7/Scripts/SDL/avatar.sdl#L181).

A few thoughts:
* The `Bone_LHand` works well as an attach point for the hand (and the KI Insert anim can be used), but I don't think there's any existing attach point for the left hip so we would need to add a new bone to the avatars.
* In some cases, an object may be cloned when being picked up. It should be fine to use the clone's key in the SDL, but I'm unsure if that has implications for users linking in after the cloning happened.
* We probably (in this simple system) don't want to allow cross-Age carrying, so we need to make sure these objects get unattached when linking out.
* For now, let's assume that objects can only be picked up/set down via some sort of clickable. No dropping things whenever you feel like.
* Only one item equipped to each location at a time.
* Again in the interest of simplicity, anims like "holding up a lantern" while an object is equipped are out of scope

**Comment by Hoikas** [comment]
I'm pretty well in agreement on these points, except I think we should add a new STATEDESC for inventory items and have an array in the avatar STATEDESC. Updating the SDL is hard, yo (especially a core one such as this...) Instead of adding bones, maybe we could store an offset point per item?

**Reply by dpogue** [reply]
Offset point from a given bone might work (although morphs might also have an impact here?)

**Comment by TheScarFr** [comment]
dpogue said:
>     We probably (in this simple system) don't want to allow cross-Age carrying, so we need to make sure these objects get unattached when linking out.

To allow cross-Age carrying could be usefull for puzzles...
Maybe it could be disallowed only in public Ages?

Another way of carrying something in the left hand could be to use the AvatarCustomization's Accessory of the Hands.
But as I don't know at all how it would technically work, I let you see...
I just wanted to write down this eventual possibility.
